csharp/ql/lib/semmle/code/csharp/security/dataflow/ZipSlipQuery.qll

```diff
@@ -3,147 +3,360 @@
+  PathCombinerToGetFullPathTaintTrackingConfiguration() {
+    this = "PathCombinerToGetFullPathTaintTrackingConfiguration "
+  }
+
+  override predicate isSource(DataFlow::Node node) {
+    exists(UnsanitizedPathCombiner pathCombiner | node = DataFlow::exprNode(pathCombiner))
+  }
+
+  override predicate isSink(DataFlow::Node node) {
+    exists(MethodCallGetFullPath mcGetFullPath |
+      node = DataFlow::exprNode(mcGetFullPath.getArgument(0))
+    )
+  }
+}
+
+private predicate safeCombineGetFullPathSequence(MethodCallGetFullPath mcGetFullPath, Expr q) {
+  // Path.Combine THEN Path.GetFullPath is applied (with possibly arbitrary mutations)
+  exists(DataFlow::Node source, PathCombinerToGetFullPathTaintTrackingConfiguration taintTracking |
+    taintTracking.hasFlow(source, DataFlow::exprNode(mcGetFullPath.getArgument(0)))
+  ) and
+  exists(GetFullPathToQualifierTaintTrackingConfiguration qualifierTaintTracker |
+    qualifierTaintTracker.hasFlow(DataFlow::exprNode(mcGetFullPath), DataFlow::exprNode(q))
+  )
+}
+
+class RootSanitizerMethodCall extends SanitizerMethodCall {
+  /*
+   * The set of /valid/ Guards of RootSanitizerMethodCall.
+   *
+   *    IN CONJUNCTION with BOTH
+   *      Path.Combine
+   *      AND Path.GetFullPath
+   *    OR
+   *      There is a direct flow from Path.GetFullPath to qualifier of RootSanitizerMethodCall.
+   *
+   *    It is not simply enough for the qualifier of String.StartsWith
+   *    to pass through Path.Combine without also passing through GetFullPath AFTER.
+   */
+
+  RootSanitizerMethodCall() {
+    exists(RootSanitizerMethod sm | this.getTarget() = sm) and
+    exists(Expr q, AbstractValue v |
+      this.getQualifier() = q and
+      v.(AbstractValues::BooleanValue).getValue() = true and
+      exists(MethodCallGetFullPath mcGetFullPath | safeCombineGetFullPathSequence(mcGetFullPath, q))
+    )
+  }
+
+  override Expr getFilePathArgument() { result = this.getQualifier() }
+}
+
+class ZipSlipGuard extends Guard {
+  /*
+   * The set of Guards of RootSanitizerMethodCall that are used IN CONJUNCTION with
+   *      Path.GetFullPath - it is not simply enough for the qualifier of String.StartsWith
+   *      to pass through Path.Combine without also passing through GetFullPath.
+   */
+
+  ZipSlipGuard() { this instanceof SanitizerMethodCall }
+
+  Expr getFilePathArgument() { result = this.(SanitizerMethodCall).getFilePathArgument() }
+}
+
+abstract class SanitizerMethodCall extends MethodCall {
+  SanitizerMethodCall() { this instanceof MethodCall }
+
+  abstract Expr getFilePathArgument();
+}
+
+class SantizedGuardTaintTrackingConfiguration extends TaintTracking2::Configuration {
+  SantizedGuardTaintTrackingConfiguration() { this = "SantizedGuardTaintTrackingConfiguration" }
+
+  override predicate isSource(DataFlow::Node source) { source instanceof DataFlow::ParameterNode }
+
+  override predicate isSink(DataFlow::Node sink) {
```

Comment:
Can maybe just make RootSanitizerMethodCall the sink and make these additional taint steps

Reply:
I'm not quite following - could you clarify?

Reply:
Actually nevermind, I had something in mind I thought would simplify things but then realized I was further complicating logic.